### ISSUE TITLE
Memory alignment check

### ISSUE THREAD

**Issue by upcFrost** [hrydgard/ppsspp#7340]
Not very sure (tested on the disassembled and modified game), but in some cases PPSSPP seems to behave differently from the real system on memory read ops. It's not actually a bug (it allows to play some faulty games), but it can be useful for debugging.

Example:
In the following memory chunk, read the 4-byte at +0x06
00 | 00 | 00 | 00 | 00 | 00 | 54 | 01 | 00 | 00 | 00 | 00 | 00 | 00 | 00 | 00
PPSSPP's result is 0x00000154, while PSP's (and JPCSP's) result'll be 0x01540000

**Comment by unknownbrackets** [comment]
Hmm, well, I doubt any compilers generate unaligned access, but yeah, we should probably align word reads if that's what the cpu does.  I guess we've never tested it.  If we're doing something differently from the cpu it's a bug, and there shouldn't be any performance hit to handling this correctly if so.

Can you submit a pspautotest (https://github.com/hrydgard/pspautotests) showing this behavior if you already have code handy?

-[Unknown]

**Comment by upcFrost** [comment]
>I doubt any compilers generate unaligned access

Sure ;) but translator's dirty hands can do. As i wrote - that's more of a debugging thing.

>if you already have code handy?

That'd be quite hard - i'm not a dev, for translation i'm using disasm, so i don't have any kind of test code (closed-source project of 1.5 Gb is not the best example). Maybe it can suffice - i've made up fast and dirty example out of text.pbp demo. 

I'm making the following change
0890046C lbu v1,0x0(v0) -> lh v1,0x0(v0)
In the original file it takes changing 00 00 43 90 to 00 00 43 84 on 0x162C

This way, 2-byte lh will become suspect to the alignment.
Results:
In PPSSPP 0.9.9.1 Stable
![ppsspp](https://cloud.githubusercontent.com/assets/287771/5790707/4476b656-9eae-11e4-98e9-e1510592604f.PNG)
In JPCSP r3681 (last nightly)
![jpcsp](https://cloud.githubusercontent.com/assets/287771/5790715/6eef5fd2-9eae-11e4-9101-f886a037a8a2.PNG)
On the real hardware - just black screen. 
Ok, the test is not fully correct, because i'm pretty sure it doesn't show any text because of lack of fonts (so it'll take some time to clean it up leaving only latin fonts). Still, if you'll look through the given code in ppsspp's and jpcsp's disasms, you'll see the same stuff i saw on psp - it takes different letters because of alignment, which's ignored in ppsspp

**Comment by unknownbrackets** [comment]
So, I just did a check on hardware.  It doesn't force alignment, it hits an exception.  So, if you want your software to work on a PSP, don't do this.

JPCSP's behavior is not crazy but it's not what hardware does.

Technically, we should probably trap the debugger on unaligned address, I guess, as the PSP does.

-[Unknown]

**Comment by CensoredUsername** [comment]
Just had a run-in with this. It was quite surprising to have software which ran on the emulator but which failed on the actual machine.

As far as I know, it traps the PSP, which causes it to halt and eventually reboot. On the Vita, it will also trap, leading to a blue screen of death with a dump of the register contents and most importantly the program counter, making it possible to debug it via a breakpoint in PPSSPP.

As far as I understand it, the PPSSPP jit emits a MOVZX on x64 when LW is encountered. As x64 does not trap on unaligned access (unless the appropriate control registers are set but this depends on the OS and as far as I know at least Windows doesn't) checking for this would require significantly more code to be generated as you'd need to check any LW op's target..

As it is something which is very uncommon in proper psp games (I'm not sure if you can even handle a bus error in the application) it is of course not a problem for just emulating original games. It would be a nice QoL fix when working on translations though.

**Comment by hrydgard** [comment]
Yeah, I don't think we need to do the check in fast memory mode which how people normally play games, but I guess it wouldn't be a bad thing to add to "slow memory" mode, so I'm leaving this issue open, of course. Not promising to work on this anytime soon though :)

**Comment by unknownbrackets** [comment]
Just as a warning: there are many other ways you can have software run in the emulator that won't run on a PSP.

More significant than alignment is the dcache, which we also don't emulate (but is harder to emulate.)  On the PSP, you might find drawing or other activity operating differently if you forget to invalidate/flush the dcache for the memory you send to the GPU.

For anyone wanting to implement this: note that we currently optimize proper unaligned loads (lwl/lwr pairs) into an lw within the jit.

-[Unknown]

**Comment by benderscruffy** [comment]
has this been implemented yet ??

**Comment by unknownbrackets** [comment]
IR checks this, as noted in #15879.  I guess we could check in interp as well, although it'd definitely make it slower.

-[Unknown]
